### PR TITLE
Fix bug where stats endpoints were calculating _all_ riak_kv stats

### PR DESCRIPTION
**Pull request by russelldb**
Since adding many more stats, and most of the infrastructure for
ad hoc querying of stats, the stat calculation code for the
(not yet legacy) endpoints was calculating all stats for riak_kv.

As there are about (ring_size \* vnode stats) + (fsm stages \* fsm stats)
more stats now, this calculation, understandbly, took a long time.

This patch instead only calculates the minimum subset of stats
needed to support the (not yet legacy) stats endpoints.

### CODE REVIEW

File: src/riak_kv_stat_bc.erl [hunk 1]
```diff
@@ -163,26 +162,43 @@ produce_stats() ->
...
 %% with commas for underscores. Uses legacy_stat_map to generate
 %% legacys stats from the new list of stats.
-backwards_compat(Stats) ->
-    [bc_stat(Old, New, Type, Stats) || {Old, New, Type} <- legacy_stat_map()].
+legacy_stats() ->
+    {Legacy, _Calculated} = lists:foldl(fun({Old, New, Type}, {Acc, Cache}) ->
+                                                bc_stat({Old, New, Type}, Acc, Cache) end,
+                                        {[], []},
+                                        legacy_stat_map()),
+    lists:reverse(Legacy).
```
Comment (seancribbs) [inline src/riak_kv_stat_bc.erl:170]: Why not `foldr` instead of `foldl` + `reverse`?

**Comment by seancribbs** [comment]
@russelldb I don't know an adequate way to test the latency improvement, but I can see how computing less would be an improvement. I left only one nitpick for you, otherwise :+1:. You might want someone more capable than I to make a before/after comparison.

File: src/riak_kv_stat_bc.erl (hunk 1 above)
Comment (russelldb) [inline src/riak_kv_stat_bc.erl:170]: Why? The `reverse` was an add on when I was eyeballing the results. There is no API guarantee of order for the stats, but adding the reverse maintains the old order. I'll make it a `foldr` instead. Looks cleaner. Thank you.

**Comment by russelldb** [comment]
I've pushed a change.

As regards testing…

This branch https://www.dropbox.com/s/km0v3086l89ts4q/kv505.png
1.3.0 https://www.dropbox.com/s/45l76mld8l3m0jv/1.3.0.png
1.2.1 https://www.dropbox.com/s/3lr17jasj036xu5/1.2.1.png

As you can see current branch is way better than 1.3.0, but worse than 1.2.1. I haven't been able to pinpoint that difference yet.

However, using this branch (https://github.com/basho/riak_core/tree/kv505-serve-from-cache) that only ever serves stats from the cache, and fetches them in the background gets this

https://www.dropbox.com/s/hcdjuzffskqma7b/kv505%2Bcache_bg.png

Which is pretty much ideal.

To reproduce results run the basho_bench configs against riak with the app.config found in this gist https://gist.github.com/russelldb/5301024b1b39ff4f35b0

**Comment by seancribbs** [comment]
Thanks! You still have my :+1:

**Comment by jrwest** [comment]
running against `verify_riak_stats` in riak test I was able to provoke an error (w/ or w/o basho/riak_core#286 ). The error does not appear against 1.3 w/o this PR (regardless of whether basho/riak_core#286 is included). not sure if this error is local to my environment:

```
webmachine error: path="/stats"
{error,{error,function_clause,[{proplists,delete,[disk,{error,{function_clause,[{riak_kv_stat_bc,'-pipe_stats/0-lc$^0/1-0-',[{riak_pipe,[{{riak_pipe,pipeline,active},0},{{riak_pipe,pipeline,create},[{count,0},{one,0}]},{{riak_pipe,pipeline,create,error},[{count,0},{one,0}]}]}],[{file,"src/riak_kv_stat_bc.erl"},{line,334}]},{riak_kv_stat_bc,pipe_stats,0,[{file,"src/riak_kv_stat_bc.erl"},{line,334}]},{riak_kv_stat_bc,produce_stats,0,[{file,"src/riak_kv_stat_bc.erl"},{line,149}]},{timer,tc,3,[{file,"timer.erl"},{line,194}]},{folsom_metrics,histogram_timed_update,4,[{file,"src/folsom_metrics.erl"},{line,177}]},{riak_core_stat_cache,'-do_get_stats/2-fun-0-',4,[{file,"src/riak_core_stat_cache.erl"},{line,182}]}]}}],[{file,"proplists.erl"},{line,362}]},{riak_kv_wm_stats,get_stats,0,[{file,"src/riak_kv_wm_stats.erl"},{line,88}]},{riak_kv_wm_stats,produce_body,2,[{file,"src/riak_kv_wm_stats.erl"},{line,77}]},{webmachine_resource,resource_call,3,[{file,"src/webmachine_resource.erl"},{line,183}]},{webmachine_resource,do,3,[{file,"src/webmachine_resource.erl"},{line,141}]},{webmachine_decision_core,resource_call,1,[{file,"src/webmachine_decision_core.erl"},{line,48}]},{webmachine_decision_core,decision,1,[{file,"src/webmachine_decision_core.erl"},{line,555}]},{webmachine_decision_core,handle_request,2,[{file,"src/webmachine_decision_core.erl"},{line,33}]}]}}
```

**Comment by russelldb** [comment]
Yeah, problem is here https://github.com/basho/riak_kv/pull/506/files#L0L322 it's fixed in amongst https://github.com/basho/riak_kv/issues/508. I should have lumped them all together. It has been a messy journey.

I've pushed a change that fixes it on this branch.

**Comment by jrwest** [comment]
confirmed that fix addressed my issue. no longer seeing that error

**Comment by jrwest** [comment]
here are my perf numbers from running on r1s02.bos ( 1 x quad core 2.4Ghz E5620 w/ 32GB RAM and 24 x 1TB 7200 RPM drives in RAID 5 (22TB usable)):

https://gist.github.com/jrwest/bc63fcd33eaed326191d

**Comment by russelldb** [comment]
Squashed and merged

**Pull request closed without merging**